migrations/20200422144952_structure_project_version_continuation_data.js

```diff
@@ -0,0 +1,156 @@
+    stringifier && stringifier.write([TYPES.LICENCE_NUMBER, projectId, versionId, projectStatus, versionStatus]);
+  }
+
+  if (ppl.length > 1) {
+    console.log(`Multple licence numbers found for versionId: ${versionId}`);
+    stringifier && stringifier.write([TYPES.MULTIPLE, projectId, versionId, projectStatus, versionStatus]);
+  }
+
+  if (!date) {
+    console.log(`Unable to parse expiry date, versionId: ${versionId}`);
+    stringifier && stringifier.write([TYPES.DATE, projectId, versionId, projectStatus, versionStatus]);
+  }
+
+  return {
+    'project-continuation': [
+      {
+        'licence-number': get(ppl, 0, null),
+        'expiry-date': date ? date.format('YYYY-MM-DD') : null
+      }
+    ]
+  }
+};
+
+exports.transform = transform;
+
+exports.up = function(knex) {
+  const stringifier = csv();
+  stringifier.write(['Type', 'Project ID', 'Version ID', 'Project Status', 'Version Status']);
+  return Promise.resolve()
+    .then(() => {
+      return knex('project_versions')
+        .select('project_versions.id')
+        .join('projects', 'project_versions.project_id', 'projects.id')
+        .where({ 'schema_version':  1 });
+    })
+    .then(versions => {
+      console.log(`found ${versions.length} versions`)
+      return versions.reduce((promise, version, index) => {
+        return promise
+          .then(() => {
+            console.log(`patching version: ${version.id}, ${index + 1} of ${versions.length}`);
+            return knex('project_versions')
+              .join('projects', 'project_versions.project_id', 'projects.id')
+              .select('project_versions.id', 'data', 'project_versions.status', 'projects.status as projectStatus', 'project_id')
+              .where({ 'project_versions.id': version.id })
+              .first()
+              .then(version => {
+                const loggingParams = {
+                  versionId: version.id,
+                  projectId: version.project_id,
+                  versionStatus: version.status,
+                  projectStatus: version.projectStatus
+                };
+                const data = transform(version.data, loggingParams, stringifier);
+                if (!data) {
+                  console.log(`Skipping ${version.id}.`)
+                  return Promise.resolve();
+                }
+                return knex('project_versions')
+                  .where({ id: version.id })
+                  .update({ data: { ...version.data, ...data } });
+              })
+              .then(() => {
+                console.log(`finshed patching version: ${version.id}, ${index + 1} of ${versions.length}`);
+              });
+          })
+          .catch(e => {
+            console.error(`Failed to update project version: ${version.id}`);
+            console.error(e.stack);
+            throw e;
+          });
+      }, Promise.resolve())
+    })
+    .then(() => {
+      stringifier.pipe(process.stdout);
```

Comment:
The pipe should be moved to the point of creation.

Reply:
I had that originally but it whinged about trying to write to a closed stream

Reply:
the close was here, in the last then

Reply:
The `end` still needs to be here, but the pipe can move up.

Reply:
I must have fucked something as it seems to work ok now. stand down.

Reply:
piping at the start results in the csv output being interrupted by console logs, and objection warnings when a field with leaves is found